sdk/nodejs/provider/server.ts

```diff
@@ -37,14 +37,26 @@ class Server implements grpc.UntypedServiceImplementation {
     engineAddr: string | undefined;
     readonly provider: Provider;
     readonly uncaughtErrors: Set<Error>;
-
-    /** Queue of construct calls. */
-    constructCallQueue = Promise.resolve();
+    private readonly _callbacks = new Map<string, any>();
 
     constructor(engineAddr: string | undefined, provider: Provider, uncaughtErrors: Set<Error>) {
         this.engineAddr = engineAddr;
         this.provider = provider;
         this.uncaughtErrors = uncaughtErrors;
+
+        // When we catch an uncaught error, we need to respond to the inflight call/construct gRPC requests
+        // with the error to avoid a hang.
+        const uncaughtHandler = (err: Error) => {
+            if (!this.uncaughtErrors.has(err)) {
+                this.uncaughtErrors.add(err);
+            }
+            // terminate the outstanding gRPC requests.
+            this._callbacks.forEach((callback) => callback(err, undefined));
+        };
+        process.on("uncaughtException", uncaughtHandler);
+        // @ts-ignore 'unhandledRejection' will almost always invoke uncaughtHandler with an Error. so
+        // just suppress the TS strictness here.
```

Comment:
What are the edge cases? Wondering for instance if it'd be better to give `uncaughtHandler` a union type or something and `if` that out with an error that'd be easier to debug if that happened?

Reply:
I replicated the existing error handler this replaces and this one here: 
https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/provider/server.ts#L641-L654

This was around for quite some time already (4+ years) and I was hesitant to change the behavior not to introduce bugs. If we'd like to change those I'd propose to do it in a separate PR in order to isolate the changes